homeworks/week4/Huy_Anh.java

```diff
@@ -0,0 +1,171 @@
+
+            //double f = generateRandomValue(intervals);
+
+            // Implemented from Mr Hien's explanation
+
+            double f = oldGenerateRandomValue(intervals);
+
+            for (int j = 0; j < length; j++) {
+                if (f >= (j + low) && f < (j + low + 1)) {
+                    count[j]++;
+                }
+            }
+        }
+
+        for (int j = 0; j < length; j++) {
+            double prob = 100*(double)count[j]/iterations;
+            if (prob > 0)
+                System.out.println("Interval [" + (int)(j + low) + "," + (int)(j + low + 1) + ") with probability " + prob + " %");
+        }
+    }
+    public static double oldGenerateRandomValue(Interval[] intervals) {
+        if (intervals == null || intervals.length == 0) return -1;
+        Random rnd = new Random();
+        double rndNumber = rnd.nextDouble(); // Uniformly generate a number
+        double firstRange = 0;
+        double span = 0;
+        for (Interval interval : intervals) // O(n) Time
+            span += interval.length();
+        for (Interval interval : intervals){ // O(n) Time
+            if(rndNumber >= firstRange/span && rndNumber < (firstRange+interval.length())/span){
+                return rndNumber*span - (firstRange - interval.low);
+            }
+            firstRange += interval.length();
+        }
+        return -1;
+    }
+    // Time Complexity: O(2kn) with k is sample times. It can be reduced to O(kn) + O(n) if taking the loop to calculate span outside the function.
+    // Space Complexity: O(1)
+
+    public static double generateRandomValue(Interval[] intervals){
+        if(intervals == null || intervals.length == 0) return -1;
+
+        //Union intervals
+        Arrays.sort(intervals, (i1, i2) -> Double.compare(i1.low, i2.low)); //Sort intervals, O(nlogn) Time
+        int count = 0; // O(1) Space
+        double low = intervals[0].low; // O(1) Space
+        double high = intervals[0].high; // O(1) Space
+        double span = 0; // O(1) Space
+        for(int i = 1; i < intervals.length; i++){ // O(n) Time
+            if(intervals[i].low > high){ //If new interval is separated
+                intervals[count] = new Interval(low, high);     // Initially, I used an ArrayList to store new interval.
+                                                                // But to reduce the space complexity, the new interval is placed on the input array
+                                                                // And count variable stores the length in new intervals (i.e the old ArrayList.size())
+                count++;
+                span += high - low;
+                low = intervals[i].low;
+                high = intervals[i].high;
+            }
+            else{
+                high = Math.max(intervals[i].high, high);
+            }
+        }
+        intervals[count] = new Interval(low, high); // Add the last union interval to list
+        span += high - low;
+        count++;
```

Comment:
Highly recommend moving this code into its own function. It's a really good practice that should become your second nature. Interviewers evaluate your SE skills by looking at how you organize your code. They can also send your code to their team members to evaluate you.

Reply:
Thank you. I was thinking of it but I thought I am not allowed to change code structure, which means I have to write my solution into prewritten function in the repl link.